File: JAVA/Program/Dpr.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Iterator;

public class Dpr
{
	private String name; 
	private String id;
	private String nama_bidang;
	private String nama_partai;
	
	public Dpr() // construct
	{
		this.name = "";
		this.id = "";
		this.nama_bidang = "";
		this.nama_partai = "";
	}
	
	public Dpr(String name, String id, String nama_bidang, String nama_partai)
	{
		this.name = name;
		this.id = id;
		this.nama_bidang = nama_bidang;
		this.nama_partai = nama_partai;
	}
	
	public void tambah(ArrayList<Dpr> list) 
	{
		Scanner scan = new Scanner(System.in); // membuat scanner
		System.out.print("Masukkan berapa anggota: ");
		int n = 0; // inisialisasi
		try {
			n = scan.nextInt(); // scanf
		} catch (Exception e) {
			System.out.println("[inputan harus bersifat integer!]");
			System.out.println();
		}
		for (int i = 0; i < n; i++) { // input masukan penambahan anggota
			System.out.println("No: " + (i + 1));
			System.out.print("Masukkan nama: ");
			String name = scan.next();
			System.out.print("Masukkan id: ");
			String id = scan.next();
			System.out.print("Masukkan nama Bidang: ");
			String nama_bidang = scan.next();
			System.out.print("Masukkan nama Partai: ");
			String nama_partai = scan.next();

			Dpr temp = new Dpr(name, id, nama_bidang, nama_partai);
			list.add(temp);
			System.out.println();
		}
	}
	
	/* getter*/
	public String getName()
	{
		return this.name;
	}
	
	public String getId()
	{
		return this.id;
	}
	
	public String getBidang()
	{
		return this.nama_bidang;
	}
	
	public String getPartai()
	{
		return this.nama_partai;
	}
	
	/* setter */
	public void setName(String name)
	{
		this.name = name;
	}
	
	public void setId(String id)
	{
		this.id= id;
	}
	
	public void setBidang(String nama_bidang)
	{
		this.nama_bidang = nama_bidang;
	}
	
	public void setPartai(String nama_partai)
	{
		this.nama_partai= nama_partai;
	}
	

	public void renameHuman(ArrayList<Dpr> list, String id, String name) // mengubah nama
	{
		for (Dpr human : list) 
		{
			if (human.getId().equals(id)) // jika id yang dicari ketemu/sama
			{
				System.out.println("Nama " + human.getName() + " telah diubah menjadi " + name);
				human.setName(name);
			}
		}
	}
	
	public void renameBidang(ArrayList<Dpr> list, String id, String bidang) // mengubah nama bidang
	{
		for (Dpr human : list) 
		{
			if (human.getId().equals(id)) // jika id yang dicari ketemu/sama
			{
				System.out.println("Nama bidang " + human.getBidang() + " telah diubah menjadi " + bidang);
				human.setBidang(bidang);
			}
		}
	}
	
	public void renamePartai(ArrayList<Dpr> list, String id, String partai) // mengubah nama partai
	{
		for (Dpr human : list) 
		{
			if (human.getId().equals(id)) // jika id yang dicari ketemu/sama
			{
				System.out.println("Nama partai: " + human.getPartai() + " telah diubah menjadi " + partai);
				human.setPartai(partai);
			}
		}
	}
	
	public void hapus(ArrayList<Dpr> list,String id) // prosedur menghpus data
	{
		Iterator<Dpr> it = list.iterator();
		while(it.hasNext())
		{
			Dpr human = it.next();
			if(human.getId().equals(id)) // jika id yang dicari ketemu /sama
			{
				System.out.println(human.getName() + " telah dihapus!");
                it.remove();
			}
		}
	}
	
	public int ketemu(ArrayList<Dpr> list,String id) // fungsi untuk menemukan id yang dicari
	{
		for(Dpr human : list)
		{
			if(human.getId().equals(id)) 
			{
				return 1; // jika ada
			}
		}
		return 0; // jika tidak ada
	}
	
	public void tampil(ArrayList<Dpr> list)  // menampilkan anggota
	{
		System.out.println("Anggota: ");
		int i = 0;
		if(list != null && !list.isEmpty()) // jika list tidak kosong
		{
			// Menghitung panjang maks dari atribut name di semua objek Human.Jika panjang 'Nama' lebih besar, maka itu panjang lebar table bagian nama.
			int maxNameLength = Math.max(list.stream().mapToInt(human -> human.getName().length()).max().getAsInt(), "Name".length());
			
			// Menghitung panjang maksimum dari atribut id di semua objek Human. id diubah menjadi string agar bisa menghitung panjangnya. Jika panjang 'Id' lebih besar, maka itu panjang lebar table bagian Id.
			int maxIdLength = Math.max(list.stream().mapToInt(human -> String.valueOf(human.getId()).length()).max().getAsInt(), "id".length());
			
			// Menghitung panjang maksimum dari atribut bidang di semua objek Human.Jika panjang 'Bidang' lebih besar, maka itu panjang lebar table bagian Bidang.
			int maxBidangLength = Math.max(list.stream().mapToInt(human -> human.getBidang().length()).max().getAsInt(), "Bidang".length());
			
			// Menghitung panjang maksimum dari atribut partai di semua objek Human. Jika panjang 'Partai' lebih besar, maka itu panjang lebar table bagian partai.
			int maxPartaiLength = Math.max(list.stream().mapToInt(human -> human.getPartai().length()).max().getAsInt(), "Partai".length());
			
			// Membuat format string untuk mencetak setiap objek Human dalam daftar. Format ini digunakan untuk mencetak tabel dengan lebar kolom yang sesuai dengan panjang maksimum dari setiap atribut.
			String format = "|| %-" + maxNameLength + "s || %-" + maxIdLength + "s || %-" + maxBidangLength + "s || %-" + maxPartaiLength + "s ||\n";

			System.out.printf(format, "Nama", "Id", "Bidang", "Partai"); // membuat tabelnya
			for (Dpr human : list)
			{
				// print setiap nama,id,bidang,partai
				System.out.printf(format, human.getName(), human.getId(), human.getBidang(), human.getPartai());
			}
			System.out.println();
		}
		else // jika list kosong
		{
			System.out.println("Kosong!");
		}
		
	}
	/* aku dapat cara function list.stream() dari gpt awkoawkoawk (emot batu) */
}

File: PYTHON/program/Dpr.py
class Dpr:
    
    __nama = ""
    __id = ""
    __nama_bidang = ""
    __nama_partai = ""
    
    # construct 
    def __init__ (self, name = "", id = "", nama_bidang = "", nama_partai = ""):
        self.__nama = name
        self.__id = id
        self.__nama_bidang = nama_bidang
        self.__nama_partai = nama_partai
    
    # getter #
    def getName(self):
        return self.__nama
        
    def getBidang(self):
        return self.__nama_bidang
    
    def getPartai(self):
        return self.__nama_partai
    
    def getId(self):
        return self.__id
        
    # setter #
    def setName(self, name):
        self.__nama = name
        
    def setBidang(self, bidang):
        self.__nama_bidang = bidang
    
    def setPartai(self, partai):
        self.__nama_partai = partai
    
    def setId(self, id):
        self.__id = id

    def tambah(self, list):
        n = 0
        try:
            n = int(input("Masukkan Berapa anggota: "))
        except ValueError:
            print("[INPUTAN HARUS BERSIFAT INTEGER!]")
            print()
        for i in range(n):
            print("No: " + str(i + 1))
            name = input("Masukkan Nama: ")
            id = input("Masukkan ID: ")
            nama_bidang = input("Masukkan Bidang: ")
            nama_partai = input("Masukkan Partai: ")
            
            temp = Dpr(name, id, nama_bidang, nama_partai)
            list.append(temp)
            print()


    def renameName(self,list, id, name): # mengubah nama
         for human in list:
             if human.getId() == id:
                 print("Nama " + human.getName() + " telah diubah menjadi " + name)
                 human.setName(name)

    def renameBidang(self,list, id, bidang): # mengubah bidang
         for human in list:
             if human.getId() == id:
                 print("Bidang " + human.getBidang() + " telah diubah menjadi " + bidang)
                 human.setBidang(bidang)
     
    def renamePartai(self,list, id, partai): # mengubah partai
         for human in list:
             if human.getId() == id:
                 print("Partai " + human.getPartai() + " telah diubah menjadi " + partai)
                 human.setPartai(partai)

    def hapus(self,list, id): # menghapus data
        for human in list[:]:
            if human.getId() == id:
                print(human.getName() +" telah dihapus!")
                list.remove(human)

    def ketemu(self,list, id): # mencari id anggota
        for human in list:
            if human.getId() == id:
                return 1
        return 0

    def tampil(self, list): # menampilkan anggota
        print("Anggota: ")
        if list and len(list) > 0:  # jika list tidak kosong
            
			# Menghitung panjang maks dari atribut name di semua objek Human.Jika panjang 'Nama' lebih besar, maka itu panjang lebar table bagian Nama.
            max_name_length = max(max(len(human.getName()) for human in list), len("Nama"))
            
            # Menghitung panjang maks dari atribut name di semua objek Human.Jika panjang 'Id' lebih besar, maka itu panjang lebar table bagian Id.
            max_id_length = max(max(len(str(human.getId())) for human in list), len("Id"))
            
            # Menghitung panjang maks dari atribut name di semua objek Human.Jika panjang 'Bidang' lebih besar, maka itu panjang lebar table bagian Bidang.
            max_bidang_length = max(max(len(human.getBidang()) for human in list), len("Bidang"))
            
            # Menghitung panjang maks dari atribut name di semua objek Human.Jika panjang 'Partai' lebih besar, maka itu panjang lebar table bagian partai.
            max_partai_length = max(max(len(human.getPartai()) for human in list), len("Partai"))
            format_string = "| {0:<" + str(max_name_length) + "} | {1:<" + str(max_id_length) + "} | {2:<" + str(max_bidang_length) + "} | {3:<" + str(max_partai_length) + "} |"

            print(format_string.format("Nama", "Id", "Bidang", "Partai")) # membuat format tabel
            for human in list:
                print(format_string.format(human.getName(), human.getId(), human.getBidang(), human.getPartai())) # print setiap nama,id,bidang,partai
            print()
        else:  # jika list kosong
            print("Kosong!")
